Augment Interpreter and Interpretation Docs with Examples and Visuals (#137)
nodestream-proj/nodestream: docs/docs/reference/interpretations.md

Source nodes are always upserted.
That means that if a node with the same key already exists, it will be updated with the new properties, relationships, and labels.
If a node with the same key does not exist, it will be created.

A source node's uniqueness is defined by the combination of its type and key. 
The `type` represents a label that is applied to the node while the `key` represents the fields that are used to uniquely identify the node relative to other nodes of the same type. 
As an example, if we have two nodes of type `Person` with the same `name` field, they will be considered the same node. 
However, if we have two nodes of type `Person` with different `name` fields, they will be considered different nodes. 
The `key` field is a dictionary where the keys represent field names, and the values can be either static values or [value providers](./value-providers.md).
As an example, if we wanted to define a source node of type `Person` with a `name` field of `John`, we would use the following interpretation:

```yaml
- type: source_node
  node_type: Person
  key:
    name: John
```

Nodestream will automatically create indexes and/or constraints for the node key fields.

> As mentioned, you can also uses value providers to define the key. 

Reviewer: ```suggestion
As mentioned, you can also use value providers to define the key. 
```

This is most often what you will do. 
As an example, if we wanted to define a source node of type `Person` with a `name` field where the value of the `name` key is defined by a field called `p_name` in your input JSON, we would use the following interpretation:

```yaml
- type: source_node
  node_type: Person
  key:

>     name: !jmespath patient_name

Reviewer: ```suggestion
    name: !jmespath p_name
```

For example, if we wanted to define a `PLAYED_IN` relationship between a source node of type `Person` and a related node of type `Movie` with a `PLAYED_IN` relationship type and a `role` property, we would use the following interpretation:

```yaml
- type: relationship
  node_type: Movie
  relationship_type: PLAYED_IN
  node_key:
    title: !jmespath movie_title
  relationship_key:
    role: !jmespath role
```

This would allow us to have multiple `PLAYED_IN` relationships between the same source node and related node as long as they have different `role` properties.

### Relationship Match Strategy

By default, related nodes are created when not present based on the supplied node type and key. However, you can change this behavior by setting the `match_strategy` field. The `match_strategy` field is a string that can be one of the following values:

* `EAGER` - Related nodes will be created when not present based on the supplied node type and key. (Default)

> * `MATCH_ONLY` - Will not create a relationship when the related node does not already exists.

Reviewer 2: ```suggestion
* `MATCH_ONLY` - Will not create a relationship when the related node does not already exist.
```

* `FUZZY` - Behaves like `MATCH ONLY`, but treats the key values as regular expressions to match on.

### Working with Many Relationships

By default, the cardinality of a relationship is one-to-one (or many; depending on the data). 

> This means that the relationship interpreation will only relate to one node per instance of the interpreation. 

Reviewer 2: ```suggestion
This means that the relationship interpretation will only relate to one node per instance of the interpretation. 
```

> Often times when working with document data, you might encounter a list of values all of which should be related to the source node. 

Reviewer 2: ```suggestion
Oftentimes when working with document data, you might encounter a list of values all of which should be related to the source node. 
```

Additionally, the `node_key` and `node_properties` fields are addressed relative to the current object in the iteration.

For example, the same interpretation as above could be written like this:

```yaml
- type: relationship
  node_type: Person
  relationship_type: HAS_CHILD
  iterate_on: !jmespath children[*]
  node_key:
    name: !jmespath name
```

| Parameter Name          	| Required? 	| Type                               	| Description                                                                                                                                                                                                                                                                                                            	|
|-------------------------	|-----------	|------------------------------------	|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------	|

> | node_type               	| Yes       	| String or ValueProvider            	| Specifies the type of the node to relationship connects to. It is a required field. When a ValueProvider is used dynamic index creation and schema introspection are not supported.                                                                                                                                    	|

Reviewer 2: ```suggestion
| node_type               	| Yes       	| String or ValueProvider            	| Specifies the type of the node a relationship connects to. It is a required field. When a ValueProvider is used dynamic index creation and schema introspection are not supported.                                                                                                                                    	|
```

> The switch interpretation allows you to define multiple interpretations that will be applied. The interpretations that will be applied is determined by the value of the `switch_on` parameter. The value of the `switch_on` parameter is a value provider that will be evaluated for each source node. The value of the value provider will be used to determine which interpretation to apply. The interpretation that will be applied is the one that has the same value as the value of the `switch_on` parameter. If no interpretation has the same value as the value of the `switch_on` parameter, the default interpretation will be applied. If no default interpretation is defined, the source node will be ignored.

Reviewer 2: ```suggestion
The switch interpretation allows you to define multiple interpretations that will be applied. The interpretation that will be applied is determined by the value of the `switch_on` parameter. The value of the `switch_on` parameter is a value provider that will be evaluated for each source node. The value of the value provider will be used to determine which interpretation to apply. The interpretation that will be applied is the one that has the same value as the value of the `switch_on` parameter. If no interpretation has the same value as the value of the `switch_on` parameter, the default interpretation will be applied. If no default interpretation is defined, the source node will be ignored.
```
